File: tcp_and_proxy/nagle.py
#!/usr/bin/python -u

# $Id$

from os import O_NONBLOCK,O_NDELAY 
from socket import socket,has_ipv6,AF_INET6,AF_INET,SOCK_STREAM,IPPROTO_TCP,TCP_CORK,error as sock_error
from sys import stdin, stdout, stderr, exit, argv
from select import select
from fcntl import fcntl,F_SETFL

def usage():
  stderr.write("\nusage: "+argv[0]+" <options> host port\n\n\
  version: $Id$\n\n\
  options:\n\
  -w max_waits\n\
   max number of times the timeout value (-t) can be reached before we send the data, default is 5 \n\
  -t seconds\n\
   timeout for a single read (in seconds), default is 0.1\n\
  -s size\n\
   packet size above which we don't buffer, default is 1024 bytes\n\n")
  exit()

# main data exchnage function
def exchange(s,max_waits,timeout,size):
  # input:
  # s - socket object
  # max_waits - max number timeout can be reached
  # timeout - timeout for a single read 
  # size - data size
  # return :
  # nothing :)

  # setting every descriptor to be non blocking 
  fcntl(s, F_SETFL, O_NONBLOCK|O_NDELAY) 
  fcntl(0, F_SETFL, O_NONBLOCK)

  s_recv = s.recv
  s_send = s.send
  write  = stdout.write
  read   = stdin.read  

  more      = 0
  waits     = 0

  while 1:
    toread,[],[] = select([0,s],[],[],timeout)
    [],towrite,[] = select([],[1,s],[],timeout)
    
    if 1 in towrite and s in toread:
      data = s_recv(4096)
      if len(data) == 0:
        s.shutdown(2)
        exit()
      else:
        write(data)

    elif 0 in toread and s in towrite: 
      if more == 0:
        
        ndata = read(4096)
        
#        if len(ndata) == 0:
#          exit()
#        else: 
#          s_send(ndata)
        
        if len(ndata) < size :
          more = 1

      else:
        ndata += read(4096)

      if ndata and len(ndata) >= size or waits >= max_waits: 
        more = 0
        waits = 0
        s_send(ndata)

    elif more == 1:
      waits += 1
      
      if waits >= max_waits:
        more = 0
        waits = 0
        s_send(ndata) 

#### main stuff ####
if __name__ == '__main__':

  if len(argv) >= 3:

    len_argv = len(argv)

    max_waits = 5
    timeout = 0.1
    size = 1024

    try:
      i = 1
      while i<len_argv:

        if argv[i] == '-w':
          max_waits = int(argv[i+1])
          i+=2
        elif argv[i] == '-t':
          timeout = float(argv[i+1])
          i+=2
        elif argv[i] == '-s':
          size = int(argv[i+1])
          i+=2
        else:
          host = argv[i]
          port = int(argv[i+1])
          i+=2

    except:
      usage()

    if ":" in host and has_ipv6 == True:
      s = socket(AF_INET6, SOCK_STREAM)
    else:
      s = socket(AF_INET, SOCK_STREAM)

#    s.setsockopt(IPPROTO_TCP, TCP_CORK,1)

    try:
      s.connect((host, port))
    except sock_error:
      stderr.write("[-] problem connecting to "+str(host)+":"+str(port)+"\n")
      s.close()
      exit()

    try:
      exchange(s,max_waits,timeout,size)
    except KeyboardInterrupt:
      pass      

    s.close()
  
  else:
    usage()

File: tcp_and_proxy/nagle.c
/*

  $Id$

*/

#include <fcntl.h>
#include <netdb.h>
#include <netinet/tcp.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/select.h>
#include <sys/socket.h>
#include <unistd.h>

#define BUFFSIZE 4096 

void die (char *mesg){
  perror(mesg);
  exit(1);
}

int main (int argc, char **argv) {

  int sock,retval,len;
  struct addrinfo hints;
  struct addrinfo *rp;
  struct timeval tv;
  fd_set rfds, wfds; 
  char* buff[BUFFSIZE];
  int state = 1;

  if (argc != 3) {
    fprintf(stderr, "usage: %s <server_ip> <port>\n",argv[0]);
    exit(1);
  }

  memset(&hints,0,sizeof(hints));
  hints.ai_family = AF_UNSPEC;    /* Allow IPv4 or IPv6 */
  hints.ai_socktype = SOCK_STREAM; 
  hints.ai_flags = 0;
  hints.ai_protocol = 0; 

  if (getaddrinfo(argv[1],argv[2],&hints,&rp) != 0)
    die("getaddrinfo()");
 
  if ((sock = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol)) < 0) 
    die("socket()");
  
  if (setsockopt(sock,IPPROTO_TCP, TCP_CORK,&state,sizeof(state)) <0)
    die("setsockopt()");
 
  if (connect(sock,rp->ai_addr,rp->ai_addrlen) == -1)
    die("connect()");

  
//  flags = fcntl(sock, F_GETFL, 0);
  fcntl(sock, F_SETFL, O_NONBLOCK);
  fcntl(0, F_SETFL, O_NONBLOCK);
  fcntl(1, F_SETFL, O_NONBLOCK);

  for(;;){

    tv.tv_sec = 30;
    tv.tv_usec = 0;

    FD_ZERO(&rfds);
    FD_ZERO(&wfds);
    FD_SET(0, &rfds);
    FD_SET(sock, &rfds);
//    FD_SET(1, &wfds);
    FD_SET(sock, &wfds);
    
    retval = select(sock+1, &rfds, &wfds, NULL, &tv);

    if (retval == -1) {

      perror("select()");

    } else if (retval) {

      if (FD_ISSET(0,&rfds) && FD_ISSET(sock,&wfds)) {

        len = read(0,buff,sizeof(buff));

        if (len < 1){
          shutdown(sock,SHUT_RDWR);
          exit(0);
        }

        send(sock,buff,len,0); 

      } 
      
      if (FD_ISSET(sock,&rfds)) {

        len = recv(sock,buff,sizeof(buff),0);

        if (len <= 0){
          shutdown(sock,SHUT_RDWR);
          exit(0);
        }

        write(1,buff,len); 

      }
    }
  }
  
  exit(0);
}
